server/src/test/java/org/elasticsearch/action/bulk/BulkOperationTests.java

```diff
@@ -663,29 +759,74 @@ public void testNodeClosureRejectsFailureStoreDocument() {
         verify(observer, times(1)).waitForNextChange(any());
     }
 
+    /**
+     * Throws an assertion error with the given message if the client operation executes
+     */
+    private static BiConsumer<BulkShardRequest, ActionListener<BulkShardResponse>> assertNoClientInteraction() {
+        return (r, l) -> fail("Should not have executed shard action on blocked cluster");
+    }
+
     /**
      * Accepts all write operations from the given request object when it is encountered in the mock shard bulk action
      */
-    private BulkShardResponse acceptAllShardWrites(BulkShardRequest request) {
-        return new BulkShardResponse(
-            request.shardId(),
-            Arrays.stream(request.items()).map(item -> requestToResponse(request.shardId(), item)).toArray(BulkItemResponse[]::new)
-        );
+    private static BiConsumer<BulkShardRequest, ActionListener<BulkShardResponse>> acceptAllShardWrites() {
+        return (BulkShardRequest request, ActionListener<BulkShardResponse> listener) -> {
+            listener.onResponse(
+                new BulkShardResponse(
+                    request.shardId(),
+                    Arrays.stream(request.items()).map(item -> requestToResponse(request.shardId(), item)).toArray(BulkItemResponse[]::new)
+                )
+            );
+        };
+    }
+
+    /**
+     * When the request is received, it is marked as failed with an exception created by the supplier
+     */
+    private BiConsumer<BulkShardRequest, ActionListener<BulkShardResponse>> failWithException(Supplier<Exception> exceptionSupplier) {
+        return (BulkShardRequest request, ActionListener<BulkShardResponse> listener) -> { listener.onFailure(exceptionSupplier.get()); };
     }
```

Comment:
Went with a supplier here so that the exception would always be a new instance, and generated closer to where it was thrown. I figured if this exception were ever to work its way into being uncaught, then the stack trace would more correctly indicate where it was being thrown from.